ui/src/tugboat/components/ImageUpload.tsx

```diff
@@ -0,0 +1,288 @@
+                        <DraggerImg src={localState.imageSrc} />
+                    ) : null}
+                    {!imageLoading && !localState.imageSrc ? (
+                        <DraggerMessage>
+                            <UploadOutlined /> Upload an Image
+                        </DraggerMessage>
+                    ) : null}
+                </Dragger>
+            </div>
+        </React.Fragment>
+    );
+};
+
+// TODO: consider making a Promise<Image> instead of callback
+// TODO: consider moving to a webworker
+export const downscaleImage = ({
+    file,
+    maxFileBytes,
+    newFileName,
+    onSuccess,
+    onError,
+}: {
+    file: File;
+    maxFileBytes: number;
+    newFileName?: string;
+    onSuccess: Function;
+    onError: Function;
+}) => {
+    const path = require('path');
+    const originalFileName = path.basename(file.name);
+    const fileName = newFileName || `${originalFileName}_c.jpg`;
+    // console.log(`Compressing ${file.name} (${file.size}Bytes)`);
+    // maxPixels captures the maximum pixels in an image that's maxFileBytes in size.
+    // 3Bytes/px was calculated from random data case of png/jpg quality 1 https://superuser.com/questions/636333/what-is-the-largest-size-of-a-640x480-jpeg
+    const maxPixels = maxFileBytes / 3;
+    const reader = new FileReader();
+    reader.onload = (event) => {
+        const img = new Image();
+        img.src = event.target ? (event.target.result as string) : '';
+        img.onload = () => {
+            const canvas = document.createElement('canvas');
+            const curPixels = img.width * img.height;
+            const scale = Math.sqrt(maxPixels) / Math.sqrt(curPixels);
+            canvas.width = img.width * scale;
+            canvas.height = img.height * scale;
+            const ctx = canvas.getContext('2d');
+            if (ctx) {
+                ctx.drawImage(img, 0, 0, canvas.width, canvas.height);
+                ctx.canvas.toBlob(
+                    (blob: Blob | null) => {
+                        if (blob) {
+                            const compressedFile = new File([blob], fileName, {
+                                type: 'image/jpeg',
+                                lastModified: Date.now(),
+                            });
+                            // console.log(`Generated ${fileName} (${compressedFile.size}Bytes)`);
+                            onSuccess(compressedFile);
+                        }
+                        onError('canvas.toBlob failed.');
+                    },
+                    'image/jpeg',
+                    1
+                );
+            } else {
+                if (onError) {
+                    onError('Invalid context.');
+                }
+            }
+        };
+        reader.onerror = (error) => {
+            Sentry.captureException(error);
+            console.log(error);
+            if (onError) {
+                onError('Error compressing image.');
```

Comment:
Consider passing the source error too, so that calling code can do something with it:

```suggestion
                onError('Error compressing image.'. error);
```

I also might consider making a specific error, and passing `error` as the cause:

```
class ImageCompressionError {
   constructor(readonly cause: Error) {
         super(`Error compressing image: ${cause}`);
   }
}

throw new ImageCompressionError(error);
```